papermill/engines.py

```diff
@@ -0,0 +1,310 @@
+    no_tqdm = False
+except OSError:
+    no_tqdm = True
+
+from .preprocess import PapermillExecutePreprocessor
+from .iorw import write_ipynb
+
+
+class PapermillEngines(object):
+    """
+    The holder which houses any engine registered with the system.
+    This object is used in a singleton manner to save and load particular
+    named EngineBase objects for reference externally.
+    """
+
+    def __init__(self):
+        self._engines = {}
+
+    def register(self, name, engine):
+        self._engines[name] = engine
+
+    def get_engine(self, name=None):
+        """
+        Retrieves an engine by name.
+        """
+        engine = self._engines.get(name)
+        if not engine:
+            raise PapermillException("No engine named '{}' found".format(name))
+        return engine
+
+    def execute_notebook_with_engine(self, engine_name, nb, kernel_name, **kwargs):
+        """
+        Fetches the approiate engine and executes the nb object against it.
+        """
+        return self.get_engine(engine_name).wrap_and_execute_notebook(nb, kernel_name, **kwargs)
+
+
+def catch_nb_assignment(func):
+    """
+    Wrapper which helps catch `nb` keyword arguments and assign onto self
+    when passed to the wrapped function.
+    Used for callback methods when the caller may optionally have a new copy
+    of the originally wrapped `nb`object.
+    """
+
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        nb = kwargs.get('nb')
+        if nb:
+            # Reassign if executing notebook object was replaced
+            self.nb = nb
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class EngineNotebookWrapper(object):
+    """
+    This class is a wrapper for notebook objects to house execution state
+    related to the notebook being run through an engine. In particular the
+    EngineNotebookWrapper provides common update callbacks for use within
+    engines to facilitate metadata and persistence actions in a shared manner.
+    """
+
+    DEFAULT_BAR_FORMAT = "{l_bar}{bar}{r_bar}"
+    PENDING = "pending"
+    RUNNING = "running"
+    COMPLETED = "completed"
+
+    def __init__(self, nb, output_path=None, log_output=False, progress_bar=True):
+        # Deep copy the input to isolate the object being executed against
+        self.nb = copy.deepcopy(nb)
+        self.output_path = output_path
+        self.log_output = log_output
+        self.set_timer()
+        self.pbar = None
+        if progress_bar and not no_tqdm:
+            self.set_pbar()
+
+    def set_pbar(self):
+        """
+        Initializes the progress bar object for this notebook run.
+
+        This is called automatically when constructed.
+        """
+        bar_format = self.DEFAULT_BAR_FORMAT
+        if self.log_output:
+            # We want to inject newlines if we're printing content between enumerations
+            bar_format += "\n"
+
+        if 'ipykernel' in sys.modules:
```

Comment:
This solves https://github.com/nteract/papermill/issues/165 too (was moving this and remembered to check if this was easy)